File: mappers-lua/57.lua
--
--
-- http://wiki.nesdev.com/w/index.php/INES_Mapper_057
--

MapperName = "Mapper 57"
MapperNumber = 57
DefaultPrgSize = 8*0x4000
DefaultChrSize = 8*0x2000

function DumpPrg(size)
	Reset();
	local banks = size / 0x4000
	for b = 0, banks-1 do		
		print("Reading PRG bank #" .. tostring(b) .. "...")
		WriteCpu(0x8800, { b * 32 })
		ReadAddPrg(0x8000, 0x4000)
	end
end

function DumpChr(size)
	Reset();
    local banks = size / 0x2000
    for b = 0, banks-1 do		
        print("Reading CHR bank #" .. tostring(b) .. "... ")
        -- TODO: Add H-byte
        WriteCpu(0x8000, {b})
        WriteCpu(0x8800, {b})
        ReadAddChr(0x0000, 0x2000)
    end
end

function EnablePrgRam(size)
	print("Warning: SRAM is not supported by this mapper")
end

File: FamicomDumper/mappers/57.cs
class Mapper57 : IMapper
{
    public string Name { get => "Mapper 57"; }
    public int Number { get => 57; }
    public int DefaultPrgSize { get => 8 * 0x4000; }
    public int DefaultChrSize { get => 8 * 0x2000; }

    public void DumpPrg(IFamicomDumperConnection dumper, List<byte> data, int size)
    {
        var banks = size / 0x4000;
        for (var bank = 0; bank < banks; bank++)
        {
            Console.Write($"Reading PRG bank #{bank}/{banks}... ");
            dumper.WriteCpu(0x8800, (byte)(bank << 5));
            data.AddRange(dumper.ReadCpu(0x8000, 0x4000));
            Console.WriteLine("OK");
        }
    }

    public void DumpChr(IFamicomDumperConnection dumper, List<byte> data, int size)
    {
        var banks = size / 0x2000;
        for (var bank = 0; bank < banks; bank++)
        {
            Console.Write($"Reading CHR bank #{bank}/{banks}... ");
            dumper.WriteCpu(0x8000, (byte)(0b10000000 | ((bank & 0b00001000) << 3)));
            dumper.WriteCpu(0x8800, (byte)(bank & 0b00000111));
            Console.WriteLine("OK");
            data.AddRange(dumper.ReadPpu(0x0000, 0x2000));
        }
    }

    public void EnablePrgRam(IFamicomDumperConnection dumper)
    {
        throw new NotSupportedException("PRG RAM is not supported by this mapper");
    }

    public MirroringType GetMirroring(IFamicomDumperConnection dumper)
    {
        return MirroringType.MapperControlled;
    }
}
